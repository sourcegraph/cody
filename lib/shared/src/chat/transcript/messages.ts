import type { ContextItem } from '../../codebase-context/messages'
import type { Message } from '../../sourcegraph-api'

import type { SerializedChatTranscript } from '.'
import type { NLSSearchDynamicFilter, NLSSearchResponse } from '../../sourcegraph-api/graphql/client'

/**
 * The list of context items (most important first) along with
 * a string label for the strategy used to obtain the context.
 * The strategy denotes both the retrieval and reranking mechanism.
 */
export type RankedContext = {
    strategy: string
    items: ContextItem[]
}

export interface ChatMessage extends Message {
    contextFiles?: ContextItem[]

    contextAlternatives?: RankedContext[]

    error?: ChatError

    /**
     * For messages composed in a rich text editor field, this is the representation of the editor
     * state that can be used to instantiate the editor to edit the message or to render the
     * message. This field's value is opaque to all but the rich editor, and it must validate and
     * version the value so that it can (1) support backward- and forward-compatibility and (2) fall
     * back to editing the text for invalid values.
     */
    editorState?: unknown

    /**
     * The model used to generate this chat message response. Not set on human messages.
     */
    model?: string

    /* The detected intent of the message */
    intent?: 'search' | 'chat' | 'edit' | 'insert' | undefined | null
    search?: ChatMessageSearch | undefined | null
}

export type ChatMessageWithSearch = ChatMessage & { search: ChatMessageSearch }

export interface ChatMessageSearch {
    query: string
    queryWithSelectedFilters?: string
    response?: NLSSearchResponse['search']
    selectedFilters?: NLSSearchDynamicFilter[]
}

// An unsafe version of the {@link ChatMessage} that has the PromptString
// replaced to a regular string for serialization @see ChatMessage
//
// Note: This is created as an interface so that the Kotlin type-gen does not
// break.
export interface SerializedChatMessage {
    contextFiles?: ContextItem[]
    error?: ChatError
    editorState?: unknown
    speaker: 'human' | 'assistant' | 'system'
    text?: string // Changed from PromptString
    model?: string
    intent?: ChatMessage['intent']
    search?: ChatMessage['search']
}

export interface ChatError {
    kind?: string
    name: string
    message: string

    // Rate-limit properties
    retryAfter?: string | null
    limit?: number
    userMessage?: string
    retryAfterDate?: Date
    retryAfterDateString?: string // same as retry after Date but JSON serializable
    retryMessage?: string
    feature?: string
    upgradeIsAvailable?: boolean

    // Prevent Error from being passed as ChatError.
    // Errors should be converted using errorToChatError.
    isChatErrorGuard: 'isChatErrorGuard'
}

export type ChatHistoryKey = `${string}-${string}`
export type AccountKeyedChatHistory = {
    [key: ChatHistoryKey]: UserLocalHistory
}

export interface UserLocalHistory {
    chat: ChatHistory
}

export interface ChatHistory {
    [chatID: string]: SerializedChatTranscript
}

/**
 * We need to specific a default event source as some commands can be
 * executed directly through VS Code where we cannot provide a custom source.
 * For example: Commands executed through the command palette, right-click menu or through keyboard shortcuts.
 */
export const DEFAULT_EVENT_SOURCE = 'editor'

export type EventSource =
    | typeof DEFAULT_EVENT_SOURCE
    | 'chat'
    | 'menu' // Cody command palette
    | 'sidebar'
    | 'code-action:explain'
    | 'code-action:document'
    | 'code-action:edit'
    | 'code-action:fix'
    | 'code-action:generate'
    | 'code-action:test'
    | 'custom-commands'
    | 'code-lens'
    | 'hover'
    | 'terminal'

/**
 * Create a mapping of all source types to numerical values, so telemetry can be recorded on `metadata`.
 */
export const EventSourceTelemetryMetadataMapping: Record<EventSource, number> = {
    [DEFAULT_EVENT_SOURCE]: 0,
    chat: 1,
    menu: 2,
    sidebar: 3,
    'code-action:explain': 4,
    'code-action:document': 5,
    'code-action:edit': 6,
    'code-action:fix': 7,
    'code-action:generate': 8,
    'code-action:test': 9,
    'custom-commands': 10,
    'code-lens': 11,
    hover: 12,
    terminal: 13,
}
/**
 * Converts an Error to a ChatError. Note that this cannot be done naively,
 * because some of the Error object's keys are typically not enumerable, and so
 * would be omitted during serialization.
 */
export function errorToChatError(error: Error): ChatError {
    return {
        isChatErrorGuard: 'isChatErrorGuard',
        ...error,
        message: error.message,
        name: error.name,
    }
}
