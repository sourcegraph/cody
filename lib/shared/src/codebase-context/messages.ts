import type { URI } from 'vscode-uri'

import type { RangeData } from '../common/range'
import type { Message } from '../sourcegraph-api'
import type { Range } from '../sourcegraph-api/graphql/client'

export type ContextFileType = 'file' | 'symbol'

/**
 * Fields that are common to any context item included in chat messages.
 */
interface ContextItemCommon {
    /**
     * The URI of the document (such as a file) where this context resides.
     */
    uri: URI

    /**
     * If only a subset of a file is included as context, the range of that subset.
     */
    range?: RangeData

    /**
     * The content, either the entire document or the range subset.
     */
    content?: string | null

    repoName?: string
    revision?: string

    /**
     * For anything other than a file or symbol, the title to display (e.g., "Terminal Output").
     */
    title?: string

    /**
     * The description of the context item used to display in mentions menu.
     */
    description?: string

    /**
     * The source of this context item.
     */
    source?: ContextItemSource

    /**
     * The token count of the item's content.
     */
    size?: number

    /**
     * Whether the item is excluded by Cody Ignore.
     */
    isIgnored?: boolean

    /**
     * Whether the content of the item is too large to be included as context.
     */
    isTooLarge?: boolean

    /**
     * If isTooLarage is true, the reason why the file was deemed too long to be included in the context.
     */
    isTooLargeReason?: string

    /**
     * The ID of the {@link ContextMentionProvider} that supplied this context item (or `undefined`
     * if from a built-in context source such as files and symbols).
     */
    provider?: string

    /**
     * Lucid icon name for the context item
     */
    icon?: string

    /**
     * Optional metadata about where this context item came from or how it was scored, which
     * can help a user or dev working on Cody understand why this item is appearing in context.
     */
    metadata?: string[]
}

/**
 * The source of this context.
 */
export enum ContextItemSource {
    /** Explicitly @-mentioned by the user in chat */
    User = 'user',

    /** From the current editor state and open tabs/documents */
    Editor = 'editor',

    /** From symf search */
    Search = 'search',

    /** In initial context */
    Initial = 'initial',

    /** Query-based context that is not added by user */
    Priority = 'priority',

    /** Remote search */
    Unified = 'unified',

    /** Selected code from the current editor */
    Selection = 'selection',

    /** Output from the terminal */
    Terminal = 'terminal',

    /** From source control history */
    History = 'history',

    /** Agentic context */
    Agentic = 'agentic',
}

/**
 * An item (such as a file or symbol) that is included as context in a chat message.
 */
export type ContextItem =
    | ContextItemFile
    | ContextItemRepository
    | ContextItemTree
    | ContextItemSymbol
    | ContextItemOpenCtx

/**
 * Context items to show by default in the chat input, or as suggestions in the chat UI.
 */
export interface DefaultContext {
    initialContext: ContextItem[]
    corpusContext: ContextItem[]
}

/**
 * A context item that represents a repository.
 */
export interface ContextItemRepository extends ContextItemCommon {
    type: 'repository'
    repoName: string
    repoID: string
    content: null
}

/**
 * A context item that represents a tree (directory).
 */
export interface ContextItemTree extends ContextItemCommon {
    type: 'tree'

    /** Only workspace root trees are supported right now. */
    isWorkspaceRoot: true

    content: null
    name: string
}

/**
 * An OpenCtx context item returned from a provider.
 */
export interface ContextItemOpenCtx extends ContextItemCommon {
    type: 'openctx'
    provider: 'openctx'
    title: string
    uri: URI
    providerUri: string
    mention?: {
        uri: string
        data?: any
        description?: string
    }
}

/**
 * A file (or a subset of a file given by a range) that is included as context in a chat message.
 */
export interface ContextItemFile extends ContextItemCommon {
    type: 'file'

    /**
     * Name of remote repository, this is how mention resolve logic checks
     * that we need to resolve this context item mention via remote search file
     */
    remoteRepositoryName?: string

    ranges?: Range[]
}

/**
 * A symbol (which is a range within a file) that is included as context in a chat message.
 */
export interface ContextItemSymbol extends ContextItemCommon {
    type: 'symbol'

    /** The name of the symbol, used for presentation only (not semantically meaningful). */
    symbolName: string

    /** The kind of symbol, used for presentation only (not semantically meaningful). */
    kind: SymbolKind

    /**
     * Name of remote repository, this is how mention resolve logic checks
     * that we need to resolve this context item mention via remote search file
     */
    remoteRepositoryName?: string
}

/** The valid kinds of a symbol. */
export type SymbolKind = 'class' | 'function' | 'method'

/** {@link ContextItem} with the `content` field set to the content. */
export type ContextItemWithContent = ContextItem & { content: string }

/**
 * A system chat message that adds a context item to the conversation.
 */
export interface ContextMessage extends Required<Message> {
    /**
     * Context messages are always "from" the human. (In the future, this could be from "system" for
     * LLMs that support that kind of message, but that `speaker` value is not currently supported
     * by the `Message` type.)
     */
    speaker: 'human'

    /**
     * The context item that this message introduces into the conversation.
     */
    file: ContextItem
}

export const GENERAL_HELP_LABEL = 'Search for a file to include, or type # for symbols...'
export const NO_SYMBOL_MATCHES_HELP_LABEL = ' (language extensions may be loading)'
export const FILE_RANGE_TOOLTIP_LABEL = 'Type a line range to include, e.g. 5-10...'
export const LARGE_FILE_WARNING_LABEL =
    'File too large. Add line range with : or use @# to choose a symbol'
export const IGNORED_FILE_WARNING_LABEL = 'File ignored by an admin setting.'
