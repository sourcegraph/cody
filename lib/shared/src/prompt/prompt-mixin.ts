import type { ChatMessage } from '../chat/transcript/messages'
import type { ChatModel } from '../models/types'
import { PromptString, ps } from './prompt-string'

/**
 * The preamble we add to the start of the last human open-end chat message that has context items.
 */
const CONTEXT_PREAMBLE = ps`You have access to the provided codebase context. `
/**
 * The preamble for preventing known models from hedging.
 */
const HEDGES_PREVENTION = ps`Answer positively without apologizing. `
/**
 * Answer guidelines for the Deep Cody model.
 */
const DEEP_CODY = ps`Give step-by-step guide for how-to and debug questions, with the updated code changes if any. Else, keep answer for general questions concise and informative. `

/**
 * Prompt mixins elaborate every prompt presented to the LLM.
 * Add a prompt mixin to prompt for cross-cutting concerns relevant to multiple commands.
 */
export class PromptMixin {
    /**
     * A list of default mixins to be prepended to the next human message.
     */
    private static mixins: PromptMixin[] = []
    private static hedging: PromptMixin = new PromptMixin(HEDGES_PREVENTION)

    /**
     * Prepends all mixins to `humanMessage`. Modifies and returns `humanMessage`.
     * Add hedging prevention prompt to specific models who need this.
     */
    public static mixInto(
        humanMessage: ChatMessage,
        modelID: ChatModel | undefined,
        newMixins: PromptMixin[] = []
    ): ChatMessage {
        const mixins = [...PromptMixin.mixins]

        // Handle hedging prevention for specific models
        const apologiticModels = ['3-5-sonnet', '3.5-sonnet']
        if (modelID && apologiticModels.some(model => modelID.includes(model))) {
            mixins.push(PromptMixin.hedging)
        }

        // Handle Deep Cody specific prompts
        const isDeepCodyEnabled = modelID?.includes('deep-cody')
        if (isDeepCodyEnabled && !newMixins.length) {
            mixins.push(new PromptMixin(DEEP_CODY))
        }

        // Add new mixins to the list of mixins to be prepended to the next human message.
        mixins.push(...newMixins)

        const prompt = PromptMixin.buildPrompt(mixins)
        return PromptMixin.mixedMessage(humanMessage, prompt, mixins, isDeepCodyEnabled)
    }

    private static buildPrompt(mixins: PromptMixin[]): PromptString {
        // Construct the prompt by joining all the mixins.
        return PromptString.join(
            mixins.map(m => m.prompt),
            ps`\n\n`
        ).trim()
    }

    private static mixedMessage(
        humanMessage: ChatMessage,
        prompt: PromptString,
        mixins: PromptMixin[],
        isDeepCodyEnabled = false
    ): ChatMessage {
        if (!mixins.length || !humanMessage.text) {
            return humanMessage
        }

        if (isDeepCodyEnabled) {
            return {
                ...humanMessage,
                text: ps`${prompt}\n\n[QUESTION]\n`.concat(humanMessage.text),
            }
        }

        // Stuff the prompt mixins at the start of the human text.
        // Note we do not reflect them in ChatMessage `text`.
        return {
            ...humanMessage,
            text: ps`${prompt}\n\nQuestion: ${humanMessage.text ?? ps``}`,
        }
    }

    public static getContextMixin(): PromptMixin {
        return new PromptMixin(CONTEXT_PREAMBLE)
    }
    /**
     * Add prompt to the list of mixins to be prepended to the next human message.
     * It gets reset each time it is used.
     */
    public static add(mixin: PromptMixin): void {
        PromptMixin.mixins.push(mixin)
    }

    /**
     * Creates a mixin with the given, fixed prompt to insert.
     */
    constructor(private readonly prompt: PromptString) {}
}

export function newPromptMixin(text: PromptString): PromptMixin {
    return new PromptMixin(text)
}
