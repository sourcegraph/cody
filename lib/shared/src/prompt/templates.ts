import type { URI } from 'vscode-uri'

import type { RangeData } from '../common/range'
import type { ActiveTextEditorDiagnostic } from '../editor'
import { displayPath } from '../editor/displayPath'
import { PromptString, ps } from './prompt-string'

export function populateCodeContextTemplate(
    code: PromptString,
    fileUri: URI,
    repoName?: PromptString,
    type: 'chat' | 'edit' = 'chat'
): PromptString {
    const template =
        type === 'edit'
            ? ps`Codebase context from file {filePath}{inRepo}:\n{text}`
            : ps`Codebase context from file {filePath}{inRepo}:\n\`\`\`{languageID}{filePathToParse}\n{text}\`\`\``

    const filePath = PromptString.fromDisplayPath(fileUri)
    return template
        .replaceAll('{inRepo}', repoName ? ps` in repository ${repoName}` : ps``)
        .replaceAll('{filePath}', filePath)
        .replaceAll('{filePathToParse}', ps`:${filePath}`)
        .replaceAll('{languageID}', PromptString.fromMarkdownCodeBlockLanguageIDForFilename(fileUri))
        .replaceAll('{text}', code)
}

const DIAGNOSTICS_CONTEXT_TEMPLATE = ps`Use the following {type} from the code snippet in the file: {filePath}:
{prefix}: {message}
Code snippet:
{code}
`

export function populateCurrentEditorDiagnosticsTemplate(
    diagnostic: ActiveTextEditorDiagnostic,
    uri: URI
): PromptString {
    const { type, message, text } = PromptString.fromTextEditorDiagnostic(diagnostic, uri)

    return DIAGNOSTICS_CONTEXT_TEMPLATE.replaceAll('{type}', type)
        .replaceAll('{filePath}', PromptString.fromDisplayPath(uri))
        .replaceAll('{prefix}', type)
        .replaceAll('{message}', message)
        .replaceAll('{languageID}', PromptString.fromMarkdownCodeBlockLanguageIDForFilename(uri))
        .replaceAll('{code}', text)
}

const COMMAND_OUTPUT_TEMPLATE = 'Here is the output returned from the terminal.\n'

export function populateTerminalOutputContextTemplate(output: string): string {
    return COMMAND_OUTPUT_TEMPLATE + output
}

const SELECTED_CODE_CONTEXT_TEMPLATE = ps`My selected code from codebase file {filePath}:\n\`\`\`\n{code}\`\`\``

export function populateCurrentSelectedCodeContextTemplate(
    code: PromptString,
    fileUri: URI,
    range?: RangeData
): PromptString {
    return SELECTED_CODE_CONTEXT_TEMPLATE.replace('{code}', code).replaceAll(
        '{filePath}',
        PromptString.fromDisplayPathLineRange(fileUri, range)
    )
}

const DIRECTORY_FILE_LIST_TEMPLATE =
    'Here is a list of files from the directory contains {fileName} in my codebase: '
const ROOT_DIRECTORY_FILE_LIST_TEMPLATE = 'Here is a list of files from the root codebase directory: '

export function populateListOfFilesContextTemplate(fileList: string, fileUri?: URI): string {
    return (
        (fileUri
            ? DIRECTORY_FILE_LIST_TEMPLATE.replace('{fileName}', displayPath(fileUri))
            : ROOT_DIRECTORY_FILE_LIST_TEMPLATE) + fileList
    )
}

export function populateContextTemplateFromText(
    templateText: PromptString,
    content: PromptString,
    fileUri: URI,
    range?: RangeData
): PromptString {
    return templateText
        .replace('{displayPath}', PromptString.fromDisplayPathLineRange(fileUri, range))
        .concat(content)
}

const FILE_IMPORTS_TEMPLATE = ps`{fileName} has imported the following: `

export function populateImportListContextTemplate(importList: PromptString, fileUri: URI): PromptString {
    return FILE_IMPORTS_TEMPLATE.replace('{fileName}', PromptString.fromDisplayPath(fileUri)).concat(
        importList
    )
}

const CODE_GENERATION_CONTEXT_TEMPLATE = ps`Below is the code from file path {filePath}. Review the code outside the XML tags to detect the functionality, formats, style, patterns, and logics in use. Then, use what you detect and reuse methods/libraries to complete and enclose completed code only inside XML tags precisely without duplicating existing implementations. Here is the code:
{precedingText}<{outputTag}></{outputTag}>{followingText}
`

export function populateCodeGenerationContextTemplate(
    precedingText: PromptString,
    followingText: PromptString,
    fileUri: URI,
    tag: PromptString
): PromptString {
    return CODE_GENERATION_CONTEXT_TEMPLATE.replaceAll('{precedingText}', precedingText)
        .replaceAll('{followingText}', followingText)
        .replaceAll('{filePath}', PromptString.fromDisplayPath(fileUri))
        .replaceAll('{outputTag}', tag)
}
