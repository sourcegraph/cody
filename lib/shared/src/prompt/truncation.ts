import type { RangeData } from '../common/range'
import { TokenCounterUtils } from '../token/counter'

import type { PromptString } from './prompt-string'

export function truncatePromptString(text: PromptString, maxTokens: number): PromptString

export function truncatePromptString(text: PromptString, maxTokens: number) {
    const encoded = TokenCounterUtils.encode(text.toString())
    return encoded.length <= maxTokens
        ? text
        : text.slice(0, TokenCounterUtils.decode(encoded.slice(0, maxTokens))?.length).trim()
}

/**
 * If text was truncated, return the truncated text and range to which it was truncated.
 * If the text is shorter than maxBytes, then return the text as-is with an undefined
 * range.
 * Note: the truncated text and range may be empty (e.g., for single-line files,
 * which should be ignored anyway).
 */
export function truncateTextNearestLine(
    text: string,
    maxBytes: number
): { truncated: string; range?: RangeData } {
    if (text.length <= maxBytes) {
        return { truncated: text }
    }
    const textMaxBytes = text.slice(0, maxBytes)
    const textMaxBytesLines = textMaxBytes.split('\n')
    if (text.at(maxBytes) === '\n' || text.at(maxBytes - 1) === '\n') {
        return {
            truncated: textMaxBytes,
            range: {
                start: { line: 0, character: 0 },
                end: { line: textMaxBytesLines.length, character: 0 },
            },
        }
    }

    const truncated = textMaxBytesLines.slice(0, -1).join('\n')
    return {
        truncated,
        range: {
            start: { line: 0, character: 0 },
            end: { line: textMaxBytesLines.length - 1, character: 0 },
        },
    }
}

/**
 * Truncates text to the given number of tokens, keeping the end of the text.
 */
export function truncateTextStart(text: string, maxTokens: number): string {
    const encoded = TokenCounterUtils.encode(text)
    return encoded.length <= maxTokens
        ? text
        : TokenCounterUtils.decode(encoded.slice(-maxTokens)).trim()
}

export function truncatePromptStringStart(text: PromptString, maxTokens: number): PromptString {
    const encoded = TokenCounterUtils.encode(text.toString())

    if (encoded.length <= maxTokens) {
        return text
    }

    // We can not create a PromptString from the tokens directly as this would be
    // considered unsafe. Instead, we use the string representation to get the updated
    // character count

    const decoded = TokenCounterUtils.decode(encoded.slice(-maxTokens))
    return text.slice(-decoded.length - 1).trim()
}
