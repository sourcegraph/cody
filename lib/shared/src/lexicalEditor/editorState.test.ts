import type { SerializedLexicalNode, SerializedRootNode } from 'lexical'
import { describe, expect, test } from 'vitest'
import { URI } from 'vscode-uri'
import { ContextItemSource } from '../codebase-context/messages'
import {
    REMOTE_DIRECTORY_PROVIDER_URI,
    REMOTE_FILE_PROVIDER_URI,
    REMOTE_REPOSITORY_PROVIDER_URI,
} from '../context/openctx/api'
import { PromptString, ps } from '../prompt/prompt-string'
import {
    editorStateFromPromptString,
    inputTextWithMappedContextChipsFromPromptEditorState,
    textContentFromSerializedLexicalNode,
} from './editorState'
import {
    FILE_MENTION_EDITOR_STATE_FIXTURE,
    GENERATE_UNIT_TEST_EDITOR_STATE_FIXTURE,
    OLD_TEXT_FILE_MENTION_EDITOR_STATE_FIXTURE,
} from './fixtures'
import type { SerializedContextItemMentionNode } from './nodes'

describe('textContentFromSerializedLexicalNode', () => {
    test('empty root', () => {
        expect(
            textContentFromSerializedLexicalNode({
                type: 'root',
                children: [],
                direction: null,
                format: 'left',
                indent: 0,
                version: 0,
            } as SerializedRootNode)
        ).toEqual('')
    })

    test('fixture from chip mentions', () => {
        expect(
            textContentFromSerializedLexicalNode(
                FILE_MENTION_EDITOR_STATE_FIXTURE.lexicalEditorState.root,
                wrapMention
            )
        ).toBe('What does <<Symbol1>> in <<file-a-1.py>> do? Also use <<README.md:2-8>>.')
    })

    test('fixture from text mentions', () => {
        expect(
            textContentFromSerializedLexicalNode(
                OLD_TEXT_FILE_MENTION_EDITOR_STATE_FIXTURE.lexicalEditorState.root,
                wrapMention
            )
        ).toBe('What does <<Symbol1>> in <<file-a-1.py>> do? Also use <<README.md:2-8>>.')
    })

    test('fixture from template', () => {
        expect(
            textContentFromSerializedLexicalNode(
                GENERATE_UNIT_TEST_EDITOR_STATE_FIXTURE.lexicalEditorState.root,
                wrapMention
            )
        ).toBe(
            'Your task is to generate a suit of multiple unit tests for the functions defined inside the <<file1.go>> file.\n\nUse the <<mention the testing framework>> framework to generate the unit tests. Follow the example tests from the <<mention an example test file>> test file. Include unit tests for the following cases: <<list test cases>>.\n\nEnsure that the unit tests cover all the edge cases and validate the expected functionality of the functions'
        )
    })
})

describe('editorStateFromPromptString', () => {
    test('converts to rich mentions', async () => {
        const input = ps`What are @${PromptString.fromDisplayPath(
            URI.file('foo.go')
        )}:3-5 and @${PromptString.fromDisplayPath(URI.file('bar.go'))} about?`
        const editorState = editorStateFromPromptString(input)
        expect(editorState.lexicalEditorState.root).toEqual<SerializedRootNode>({
            children: [
                {
                    children: [
                        {
                            detail: 0,
                            format: 0,
                            mode: 'normal',
                            style: '',
                            text: 'What are ',
                            type: 'text',
                            version: 1,
                        },
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'file',
                                uri: 'file:///foo.go',
                                content: undefined,
                                source: ContextItemSource.User,
                                range: {
                                    start: {
                                        line: 2,
                                        character: 0,
                                    },
                                    end: {
                                        line: 5,
                                        character: 0,
                                    },
                                },
                            },
                            isFromInitialContext: false,
                            text: 'foo.go:3-5',
                        } satisfies SerializedContextItemMentionNode,
                        {
                            detail: 0,
                            format: 0,
                            mode: 'normal',
                            style: '',
                            text: ' and ',
                            type: 'text',
                            version: 1,
                        },
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'file',
                                uri: 'file:///bar.go',
                                content: undefined,
                                range: undefined,
                                source: ContextItemSource.Editor,
                            },
                            isFromInitialContext: false,
                            text: 'bar.go',
                        } satisfies SerializedContextItemMentionNode,
                        {
                            detail: 0,
                            format: 0,
                            mode: 'normal',
                            style: '',
                            text: ' about?',
                            type: 'text',
                            version: 1,
                        },
                    ],
                    direction: null,
                    format: '',
                    indent: 0,
                    type: 'paragraph',
                    version: 1,
                } as SerializedLexicalNode,
            ],
            direction: null,
            format: '',
            indent: 0,
            type: 'root',
            version: 1,
        })
        expect(
            textContentFromSerializedLexicalNode(editorState.lexicalEditorState.root, wrapMention)
        ).toBe('What are <<foo.go:3-5>> and <<bar.go>> about?')
    })

    test('parse templates', () => {
        const input = ps`Generate tests for @${PromptString.fromDisplayPath(
            URI.file('foo.go')
        )} using {{mention framework}} framework to generate the unit tests`
        const editorState = editorStateFromPromptString(input, {
            parseTemplates: true,
        })
        expect(editorState.lexicalEditorState.root).matchSnapshot()
        expect(
            textContentFromSerializedLexicalNode(editorState.lexicalEditorState.root, wrapMention)
        ).toBe(
            'Generate tests for <<foo.go>> using <<mention framework>> framework to generate the unit tests'
        )
    })
})

describe('inputTextWithMappedContextChipsFromPromptEditorState', () => {
    test('maps all expected mentions', async () => {
        const andNode = {
            detail: 0,
            format: 0,
            mode: 'normal',
            style: '',
            text: ' and ',
            type: 'text',
            version: 1,
        }
        const editorState = editorStateFromPromptString(ps``)
        editorState.lexicalEditorState.root = {
            children: [
                {
                    children: [
                        {
                            detail: 0,
                            format: 0,
                            mode: 'normal',
                            style: '',
                            text: 'What are ',
                            type: 'text',
                            version: 1,
                        },
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'file',
                                uri: 'file:///foo.go',
                                content: undefined,
                                source: ContextItemSource.User,
                                range: {
                                    start: {
                                        line: 2,
                                        character: 0,
                                    },
                                    end: {
                                        line: 5,
                                        character: 0,
                                    },
                                },
                            },
                            isFromInitialContext: false,
                            text: 'foo.go:3-5',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'repository',
                                uri: 'file:///bar.go',
                                content: undefined,
                                repoName: 'sourcegraph/sourcegraph',
                                repoID: '',
                            },
                            isFromInitialContext: false,
                            text: 'sourcegraph/sourcegraph',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'tree',
                                uri: 'file:///',
                                content: undefined,
                                repoName: 'sourcegraph/sourcegraph',
                                isWorkspaceRoot: true,
                                name: 'root',
                            },
                            isFromInitialContext: false,
                            text: 'myrootdir',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'symbol',
                                uri: 'symbol://mysymbol',
                                content: undefined,
                                repoName: 'sourcegraph/sourcegraph',
                                symbolName: 'mysymbol',
                                kind: 'class',
                            },
                            isFromInitialContext: false,
                            text: 'mysymbol',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'openctx',
                                title: 'myprovider',
                                provider: 'openctx',
                                uri: 'placeholder',
                                content: undefined,
                                providerUri: REMOTE_REPOSITORY_PROVIDER_URI,
                            },
                            isFromInitialContext: false,
                            text: 'myremoterepo',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'openctx',
                                title: 'myprovider',
                                provider: 'openctx',
                                uri: 'placeholder',
                                content: undefined,
                                providerUri: REMOTE_FILE_PROVIDER_URI,
                            },
                            isFromInitialContext: false,
                            text: 'myremoterepo',
                        } satisfies SerializedContextItemMentionNode,
                        andNode,
                        {
                            type: 'contextItemMention',
                            version: 1,
                            contextItem: {
                                type: 'openctx',
                                title: 'myprovider',
                                provider: 'openctx',
                                uri: 'placeholder',
                                content: undefined,
                                providerUri: REMOTE_DIRECTORY_PROVIDER_URI,
                            },
                            isFromInitialContext: false,
                            text: 'myremotedir',
                        } satisfies SerializedContextItemMentionNode,
                    ],
                    direction: null,
                    format: '',
                    indent: 0,
                    type: 'paragraph',
                    version: 1,
                } as SerializedLexicalNode,
            ],
            direction: null,
            format: '',
            indent: 0,
            type: 'root',
            version: 1,
        }
        expect(inputTextWithMappedContextChipsFromPromptEditorState(editorState)).toBe(
            'What are @file and @repo and @dir and @symbol and @repo and @file and @dir'
        )
    })
})

function wrapMention(text: string): string | undefined {
    return `<<${text}>>`
}
