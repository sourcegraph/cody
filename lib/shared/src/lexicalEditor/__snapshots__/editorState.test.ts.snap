// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`editorStateFromPromptString > parse templates 1`] = `
{
  "children": [
    {
      "children": [
        {
          "detail": 0,
          "format": 0,
          "mode": "normal",
          "style": "",
          "text": "Generate tests for ",
          "type": "text",
          "version": 1,
        },
        {
          "contextItem": {
            "content": undefined,
            "range": undefined,
            "source": "editor",
            "type": "file",
            "uri": "file:///foo.go",
          },
          "isFromInitialContext": false,
          "text": "foo.go",
          "type": "contextItemMention",
          "version": 1,
        },
        {
          "detail": 0,
          "format": 0,
          "mode": "normal",
          "style": "",
          "text": " using ",
          "type": "text",
          "version": 1,
        },
        {
          "templateInput": {
            "placeholder": "mention framework",
          },
          "type": "templateInput",
          "version": 1,
        },
        {
          "detail": 0,
          "format": 0,
          "mode": "normal",
          "style": "",
          "text": " framework to generate the unit tests",
          "type": "text",
          "version": 1,
        },
      ],
      "direction": null,
      "format": "",
      "indent": 0,
      "type": "paragraph",
      "version": 1,
    },
  ],
  "direction": null,
  "format": "",
  "indent": 0,
  "type": "root",
  "version": 1,
}
`;
